Move jetpack_file_data to wp_site_meta table in multisites

igmoweb opened the pull request:
Fixes #5763

#### Changes proposed in this Pull Request:
New function: `Jetpack_Options::is_network_option( $option_name )` that checks if an option must be saved in wp_sitemeta when multisite is installed.
Moved `jetpack_file_data` option to wp_sitemeta table.

#### Testing instructions:
Network activate Jetpack. `jetpack_file_data` should be stored now in wp_sitemeta table.

#### Proposed changelog entry for your changes:
- Enhance: Moved jetpack_file_data option to wp_sitemeta table in multisites.

zinigor commented on class.jetpack-options.php:
```diff
@@ -134,10 +156,21 @@ public static function get_option( $name, $default = false ) {
 
 		if ( $value === false && $default !== false ) {
-			update_option( $name, $default );
+			if ( $is_network_option ) {
```
I'm not sure, but I think you have this conditional backwards. Shouldn't this be `update_site_option` instead?

igmoweb replied:
Fixed.

zinigor commented on class.jetpack-options.php:
```diff
@@ -151,7 +184,14 @@ private static function update_grouped_option( $group, $name, $value ) {
 		$options[ $name ] = $value;
 
-		return update_option( self::$grouped_options[ $group ], $options );
+		if ( self::is_network_option( $name ) ) {
```
Same problem with the conditional here as with the one above.

igmoweb commented on class.jetpack-options.php:
```diff
@@ -153,13 +183,28 @@ public static function get_option( $name, $default = false ) {
 	 */
 	public static function get_option_and_ensure_autoload( $name, $default ) {
-		$value = get_option( $name );
+		// In this function the name is not adjusted by prefixing jetpack_
```
There was another error here. In this function, $name is already prefixed with 'jetpack_' so `is_network_option()` was always returning false. I used a regexp to remove it  and check against `is_network_option()`

georgestephanis replied:
Probably should indicate in the various functions in the phpdoc whether it expects the prefixed or unprefixed option names.  Possibly evaluate changing the ones using prefixed versinos to private member functions so there's a consistent usage for all public-facing methods?

igmoweb commented:
I updated the PR with some fixes and I added a few unit tests for multisite.

About making the code more DRY: I thought that at first but I'm not totally sure about secondary effects and that's more a refactoring than a little change. Basically what I did is replace `get/update/delete_option` for a check (`is_network_option()`) and then use the right function (`get_option`/`get_site_option`). I believe that by making the code DRY would end up in other problems and that it needs another different Pull Request for that but just my opinion.

zinigor commented:
Changes look good, thank you! But still I'd like to have someone else look over the code. @georgestephanis do you maybe have some suggestions on how to make the code more DRY without getting too involved? Also perhaps you know something we're forgetting here.

igmoweb commented:
Once thing I'm concerned about is that it would be nice if the old option is removed from any blog in a multisite as it won't be needed any longer. I can update the PR but I wanted to ask first. I guess there's a function to update the DB version option or similar, isn't it? It would just be to update the DB version and then clean that option.

samhotchkiss commented:
Bumping this to 4.6 to give @georgestephanis a chance to weigh in once he's back.

georgestephanis commented:
Ack, sorry, GitHub notifications hate me.  Looking now.

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -123,6 +123,30 @@ public static function is_valid( $name, $group = null ) {
+	 *
+	 * @return array
+	 */
+	public static function get_network_options() {
```
I'd sooner see this as a parameter passed to `Jetpack_Options::get_option_names()` to better keep all our options clustered for readability.

igmoweb replied:
I changed this in my last commit. I've removed ```get_network_options()``` and add a third case inside ```Jeptack_Options::get_option_names()``` so you can get the network options list with ```Jetpack_Options::get_option_names('network')```

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -132,7 +156,13 @@ public static function is_valid( $name, $group = null ) {
+			if ( self::is_network_option( $name ) ) {
+				return get_site_option( "jetpack_$name", $default );
+			}
+			else {
```
This should be on the prior line -- core coding style is `} else {`

georgestephanis replied:
Also, as the prior conditional has a return statement, this really doesn't need to be in an `else` at all.  It could just have the prior if as a short circuit.

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -153,13 +183,28 @@ public static function get_option( $name, $default = false ) {
 	 * @param string $name Option name
```
We should probably be more clear in the phpdoc if it's the raw option name, or the arg passed into the class (unprefixed)

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -153,13 +183,28 @@ public static function get_option( $name, $default = false ) {
+		if ( $is_network_option ) {
+			$value = get_site_option( $name );
+		}
+		else {
```
Core coding style should have this as `} else {` but I could see it as a ternary.  Either way.

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -153,13 +183,28 @@ public static function get_option( $name, $default = false ) {
+			if ( $is_network_option ) {
+				update_site_option( $name, $default );
+			}
+			else {
```
Core coding style should have this as `} else {`

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -173,7 +218,14 @@ private static function update_grouped_option( $group, $name, $value ) {
+		if ( self::is_network_option( $name ) ) {
+			return update_site_option( self::$grouped_options[ $group ], $options );
+		}
+		else {
```
As prior condition is doing a `return` this doesn't need to be in an `else`

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -196,7 +248,13 @@ public static function update_option( $name, $value, $autoload = null ) {
+			if ( self::is_network_option( $name ) ) {
+				return update_site_option( "jetpack_$name", $value );
+			}
+			else {
```
As prior condition is doing a `return` this doesn't need to be in an `else`

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -173,7 +218,14 @@ private static function update_grouped_option( $group, $name, $value ) {
 		$options[ $name ] = $value;
 
-		return update_option( self::$grouped_options[ $group ], $options );
+		if ( self::is_network_option( $name ) ) {
```
Will this ever be relevant for grouped options?  Or is it just a case of being thorough and building out future infastructure?

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -246,9 +304,17 @@ public static function delete_option( $names ) {
-			if ( ! delete_option( "jetpack_$name" ) ) {
-				$result = false;
+			if ( self::is_network_option( $name ) ) {
+				if ( ! delete_site_option( "jetpack_$name" ) ) {
```
Could these lines be simplified as `$result = delete_site_option( "jetpack_$name" );` ?

[`delete_site_option()`](https://github.com/WordPress/WordPress/blob/9d7ea04936125041b224b227f48ad4194066b314/wp-includes/option.php#L1057) always returns a bool anyways.

igmoweb replied:
Fixed.

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -246,9 +304,17 @@ public static function delete_option( $names ) {
+					$result = false;
+				}
 			}
+			else {
```
Should be `} else {`

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -246,9 +304,17 @@ public static function delete_option( $names ) {
+				}
 			}
+			else {
+				if ( ! delete_option( "jetpack_$name" ) ) {
```
Same as above for simplifying the conditional

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -261,7 +327,13 @@ public static function delete_option( $names ) {
+		if ( self::is_network_option( $name ) ) {
+			$options = get_site_option( self::$grouped_options[ $group ] );
+		}
+		else {
```
Should be on previous line -- `} else {`

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -261,7 +327,13 @@ public static function delete_option( $names ) {
 
 	private static function get_grouped_option( $group, $name, $default ) {
-		$options = get_option( self::$grouped_options[ $group ] );
+		if ( self::is_network_option( $name ) ) {
```
Should we even be messing around with grouped options in this PR?

igmoweb replied:
I think I don't get the point of your comment here. I'm not messing with grouped options but moving that line.

Instead of ```$options = get_option( self::$grouped_options[ $group ] );```
I use ```$options = get_site_option( self::$grouped_options[ $group ] );```
or the original one ```$options = get_option( self::$grouped_options[ $group ] );``` if it's not a multisite

georgestephanis replied:
So, my point here is that the `file_data` option we're changing is a non-grouped option, so any changes to enable network storage of grouped option won't have any bearing on this use case.

I can understand potentially adding it for API Consistency or the like, but handling grouped options may be tricky, as  it may be desirable to flag some of the options as network and some not -- so for the moment I'd prefer not to add any functionality to let grouped options be flagged as network storage.

Does that make sense?

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -270,15 +342,28 @@ private static function get_grouped_option( $group, $name, $default ) {
+		if ( $is_network_option ) {
+			$options = get_site_option( self::$grouped_options[ $group ], array() );
+		}
+		else {
```
Should be on prior line `} else {`

igmoweb replied:
Fixed

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -270,15 +342,28 @@ private static function get_grouped_option( $group, $name, $default ) {
 
 	private static function delete_grouped_option( $group, $names ) {
-		$options = get_option( self::$grouped_options[ $group ], array() );
+		$is_network_option = self::is_network_option( self::$grouped_options[ $group ] );
```
Again, should this even be messing with grouped options?  `file_data` is a non-compact option regardless.

igmoweb replied:
See my comment above

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -270,15 +342,28 @@ private static function get_grouped_option( $group, $name, $default ) {
+			if ( $is_network_option ) {
+				return update_site_option( self::$grouped_options[ $group ], $options );
+			}
+			else {
```
Should be on prior line.

igmoweb commented:
@georgestephanis 

Thanks for the comments. You're not over picky :)

1. I fixed code styling
2. I removed ```Jetpack_options:: get_network_options()``` and use ```Jetpack_Options:::get_option_names( 'network' )``` instead.
3. I don't understand well why I am messing with grouped options. I just moved the code so the grouped option is retrieved by ```get_site_option``` or ```get_option``` depending if we're in multisite or not. Maybe this is not needed at all for my changes but I did it to prevent possible breaks. You have a better big picture than me so maybe I'm missing something about grouped options.

Thanks.

igmoweb commented:
@georgestephanis 

So I removed the checks for network options in grouped options methods. See commit here: https://github.com/Automattic/jetpack/pull/5764/commits/2f4ce48a6e4b6f25746f8a9bca6fde53242035b8

And added some phpdocs in this one https://github.com/Automattic/jetpack/pull/5764/commits/44d1ed5f6623708f45ee8f4d0691c73e691246fb

What I couldn't do is to move public methods that are using prefixed options to public member methods as they are used a lot throughout the code and unit tests. I believe this change would be good for another PR so we would keep this one small.

That said. What if we delete file_data option from any blog in a multisite during the next upgrade?

Something like this would help:

```
if ( is_multisite() ) {
  delete_option( 'jetpack_file_data' );
}
```

Thing is, I cannot find any method/function in charge of upgradings. I see many functions attached to 'updating_jetpack_version' but I can't see the best place to do it.

Thanks.

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -62,6 +62,11 @@ public static function get_option_names( $type = 'compact' ) {
+
+		case 'network' :
+			return array(
+				'file_data'                     // (array) List pf absolute paths to all Jetpack modules
```
Typo `pf` => `of`

georgestephanis commented on class.jetpack-options.php:
```diff
@@ -123,15 +128,33 @@ public static function is_valid( $name, $group = null ) {
 	/**
+	 * Checks if an option must be saved for the whole network in WP Multisite
+	 *
+	 * @param string $option_name Option name. It must come _without_ jetpack_% prefix. The method will prefix the option name.
```
May want to put backticks around the `jetpack_%` bit -- the underscores on the prior word can be a touch confusing for the eye.  Same remark for subsequent uses.

igmoweb commented:
@georgestephanis I think we are fine now :)

Thanks.

georgestephanis commented:
@jeherve I don't think the changes merit a redo of the review -- it seems to have just been merging master back in to the branch?

jeherve commented:
@georgestephanis Yes indeed, my bad, I didn't check the commits.

igmoweb commented:
@georgestephanis @jeherve My bad. I didn't push my latest changes. Now is fine.

georgestephanis commented:
Just leaving a note for myself here to consider short-circuiting the `is_network_option` check when syncing to wpcom -- we may not want that there.
